test/ring/middleware/test/json.clj

```diff
@@ -337,6 +338,25 @@
       (is (= (get-in response [:headers "Content-Type"]) "application/json; some-param=some-value"))
       (is (= (:body response) "{\"foo\":\"bar\"}")))))
 
+(defn- streamable->string [body]
+  (let [baos (java.io.ByteArrayOutputStream.)]
+    (write-body-to-stream body nil baos)
+    (.toString baos "utf-8")))
+
+(deftest test-json-response-streaming
+  (testing "streaming vector body"
+    (let [handler  (constantly {:status 200 :headers {} :body [:foo :bar]})
+          response ((wrap-json-response handler {:stream? true}) {})]
+      (is (= (get-in response [:headers "Content-Type"]) "application/json; charset=utf-8"))
+      (is (= (streamable->string (:body response)) "[\"foo\",\"bar\"]"))))
+
+  (testing "streaming map body with options"
+    (let [handler  (constantly {:status 200 :headers {} :body {:foo "bar" :baz "quz"}})
+          response ((wrap-json-response handler {:stream? true :pretty true}) {})]
+      (is (let [body (streamable->string (:body response))]
```

Comment:
Sorry for taking a while to review this. It all looks good, except that this `is` clause isn't going to give good feedback if it fails. Instead, what about something like:

```clojure
  (testing "streaming map body with options"
    (let [data     (sorted-map :foo "bar" :baz "quz")
          handler  (constantly {:status 200 :headers {} :body data})
          response ((wrap-json-response handler {:stream? true :pretty true}) {})]
      (is (= (streamable->string (:body response))
             "{\n  \"baz\" : \"quz\",\n  \"foo\" : \"bar\"\n}"))))
```

Reply:
The way I was stacking the let forms was super-clunky and I've fixed that so we can see the wood for the trees.

You've previously hedged your bets in this file regarding attribute render order determinism - I was just following that pattern with the or clause. It's consistent with "JSON options" test in test-json-response and test-json-response-cps. I'd suggest if you feel confident in a deterministic attribute render order then we should fix it across the whole file in a separate PR?

Reply:
Maybe if we're doing string comparisons we should be using cheshire.core/generate-string to generate the expected cases - that way we're super clear that we're just testing the plumbing rather than making assumptions about the rendering.

Reply:
I didn't realize that other tests used `or` as well. Let's keep it consistent, then.